R/with-cache.R

```diff
@@ -0,0 +1,522 @@
+#'   represented in the cache key.
+#'
+#'   In general, `key` should use the same reactive inputs as `value`, but the
+#'   computation should be simpler. If there are other (non-reactive) values
+#'   that are consumed, such as external data sources, they should be used in
+#'   the `key` as well. Note that if the `key` is large, it can make sense to do
+#'   some sort of reduction on it so that the serialization and hashing of the
+#'   cache key is not too expensive.
+#'
+#'   Remember that the key is _reactive_, so it is not re-executed every single
+#'   time that someone accesses the cached reactive. It is only re-executed if
+#'   it has been invalidated by one of the reactives it depends on.. For
+#'   example, suppose we have this cached reactive:
+#'
+#'   ```
+#'   r <- cachedReactive(
+#'     reactive({ input$x + input$y }),
+#'     input$x, input$y
+#'   )
+#'   ```
+#'
+#' In this case, the key expression is essentially `reactive(list(input$x,
+#' input$y))` (there's a bit more to it, but that's a good enough
+#' approximation). The first time `r()` is called, it executes the key, then
+#' fails to find it in the cache, so it executes the value expression, `{
+#' input$x + input$y }`. If `r()` is called again, then it does not need to
+#' re-execute the key expression, because it has not been invalidated via a
+#' change to `input$x` or `input$y`; it simply returns the previous value.
+#' However, if `input$x` or `input$y` changes, then the reactive expression will
+#' be invalidated, and the next time that someone calls `r()`, the key
+#' expression will need to be re-executed.
+#'
+#' Note that if the cached reactive is passed to [withEvent()], then the key
+#' expression will no longer be reactive; instead, the event expression will
+#' be reactive.
+#'
+#'
+#' @section Async with cached reactives:
+#'
+#'  With a cached reactive expression, the key and/or value expression can be
+#'  _asynchronous_. In other words, they can be promises -- not regular R
+#'  promises, but rather objects provided by the \pkg{promises}  package, which
+#'  are similar to promises in JavaScript. (See [promises::promise()] for more
+#'  information.) You can also use [future::future()] objects to run code in a
+#'  separate process or even on a remote machine.
+#'
+#'  If the value returns a promise, then anything that consumes the cached
+#'  reactive must expect it to return a promise.
+#'
+#'  Similarly, if the key is a promise (in other words, if it is asynchronous),
+#'  then the entire cached reactive must be asynchronous, since the key must be
+#'  computed asynchronously before it knows whether to compute the value or the
+#'  value is retrieved from the cache. Anything that consumes the cached
+#'  reactive must therefore expect it to return a promise.
+#'
+#'
+#' @section Cache scope:
+#'
+#'   By default, a cached reactive is scoped to the running application. That
```

Comment:
I don't know how important this is, but I was left wondering if there's one overall cache, or one cache per reactive (i.e. is the bound on size per object or per app?)